rfd/0002/README.mdx

```diff
@@ -0,0 +1,198 @@
+—--
+authors: Nicolas Joseph <nicolas@nicolasjoseph.com>
+title: Ecosystem Skeleton Architecture
+state: discussion
+—--
+
+## Framework Problem Statement:
+
+We are building an opinionated framework to enabled engineers that are building data infrastructure and pipeline to focus on the business logic and not the infrastructure. We aim at providing them with the same user experience as a front end engineer building a web application. We want to provide a framework that is easy to use, easy to maintain, easy to extend and most importantly, gives the user feedback on a code iteration immediately.
+
+## Principles
+
+- **Local first**: We optimize for the user experience first and we will figure out how to deploy to production later.
+- **Easy to use**: We want to provide a framework that is easy to use, easy to maintain, easy to extend and most importantly, gives the user feedback on a code iteration immediately.
+- **Easy to test**: We want to provide a framework that has built in unit testing from the get go.
+- **Performance**: We want to provide a framework that is fast and can scale to large data sets and volumes for real life use cases.
+- **Extensible**: We want to provide a framework that is easy to extend and that can be used to build a wide range of data applications.
+
+## Architecture
+
+### Overview
+
+- Data Ingest SDK: The data ingest SDK is a simple library that can be embedded in the user's application to send data to an igloo deployment. It is typed with the User's Business Objects to ensure that the data sent to igloo is valid.
```

Comment:
Are you thinking at all about extraction? The real HOW to get data out of like shitforce and start converting it to streaming.

Reply:
That's a good callout. No, I don't think we have been thinking too much about that.

Knee jerk reaction is: 2 use cases for extraction
* pull based: we could run a function the user would implement on a cron. The function would use the before mentioned SDK to produce the events that are ingested, reading the data from the target platform apis.
* push based: the platform to extract data from has a webhook feature, and we allow that to point to igloo server for ingest - not sure how to think about types in this case.

Reply:
Yea, nexla/fivetran do it where they use REST as much as they can and read records then shit them over into events/single records. It gets slow as heck though when you want to process 2m records a day from say gcs with lots of small files...  Bunch to think through in just the data acquisition part, especially since the majority of decisions folks would want to make around data and do ml with lay outside their product engineering orgs.

Reply:
Can be an interesting area to think about partnering with folks as well as opposed to building. We might want to look into CDC providers like debezium for the DB related extractions

Reply:
@tg339 you're also just thinking about DB's which is significantly easier then business applications which want to horde data.  That's where you can really get analytics unlocks.  I think all the normal B2C, and general martech use cases are super easy compared to end to end business analytics.

Reply:
@mcginleyr1 good point. We need to do some more digging here. Feels like this should be solvable with reasonable lift. If we need to go down the full blown connector path, that's where the lift might become to heavy to justify the use cases

Reply:
It might be easier to put together toolkits so folks can build their own connectors really easily. since all they have to do is read some crappy endpoints or connections and spew the data into your server.

Reply:
Yeah - I will put together another rfd to look at what this could look like. I was looking at airbyte and how they do connectors, we might be able to do a destination into igloo and leverage their ecosystem of connectors. As well as provide low level sdk if someone wanted to build the integration directly.